go/libraries/doltcore/remotesrv/interceptors.go

```diff
@@ -34,12 +35,13 @@ type RequestCredentials struct {
 }
 
 type ServerInterceptor struct {
-	Lgr           *logrus.Entry
-	Authenticator Authenticator
+	Lgr              *logrus.Entry
+	AccessController AccessControl
 }
 
-type Authenticator interface {
-	Authenticate(creds *RequestCredentials) bool
+type AccessControl interface {
+	ApiAuthenticate(creds *RequestCredentials, lgr *logrus.Entry) *sql.Context
```

Comment:
Structurally / architecturally, it seems somewhat weird for this package to take a dependency on `sql.Context`.

Two possible alternatives:

1) Authenticate could return an `Authorizer interface` implementation, would have a single method, like `Authorize` or `ApiAuthorize` or something, which would take a description of the request (nothing for now?) and return `error` or whatever. Then the implementation for the sql server could implement that to keep the MySQLDb and the sql.Context around and perform the check.

In code, something like:

```go
type AccessControl interface {
    ApiAuthenticate(creds *RequestCredentials, lgr *logrus.Entry) Authenicator
}

type Authorizer interface {
    Authorize() bool
}
```

or something like that.

2) Could just have `ApiAuthenticate` return a new `context.Context`, instead of a `*sql.Context`. The semantics could also be that the new context became the request context – like one would expect with an interceptor chain which is annotating request-specific information (tracing, log level, authentication middleware, etc., all very typical like this...). An implementor could thread whatever state they wanted (a *sql.Context implementation, for example) through on a `context.WithValue()` implementation. If the contract became the typical middleware contract, implementations would want to change to actually thread the new context through, since it should become the new request context.

Thinking about it further, even `RequestCredentials` seems a bit off to me here, at least in the inteceptor's interface implementation and everything...just something like:

```go
type AccessControl interface {
    Authenticate(ctx context.Context) (context.Context, error)
    Authorize(ctx context.Context) error
}
```

With it up to the implementation to either slice up in the context in they way that lets it make sense, or to install middleware so that the implementation has access to what it needs...seems the most idiomatic and the most flexible way going forward. (As an example, in the gRPC context, the server `context.Context` implementation already supports `grpc.Method(ctx)` for extracting the method being called, which would come in handy for the incoming distinction between reads with CLONE_ADMIN and writes with ADMIN / SUPER or whatever the GRANT is...)

If we really care about keeping parsing of the Basic header out of the `Authenticate` implementation, an interceptor which annotated the context with basic auth details would maybe be reasonable...here, the coupling of `remotesrv` package with the default username/password pair `"root"/""` also seems quite off...